linode/manager
# feat: [M3-6785] - Firewall-NodeBalancer → Dev

## Description 📝
This PR merges the `Firewall-NodeBalancer` feature branch into the `develop` branch.

## Major Changes 🔄
- Updated the following flows to allow users to attach/assign `NodeBalancers` or `Linodes` to `Firewalls`:
  - Create A Linode
  - Create A Firewall
  - Create A NodeBalancer
- Updated the tabs for Linode, Firewall, and NodeBalancer.
- Updated events landing page to include NodeBalancer details.

## How to test 🧪

### Prerequisites
- Turn on the Firewall-Nodebalancer feature flag in dev tools
- To go through the workflow, have atleast:
   - 2 Linodes (without an assigned Firewall)
      - 'Linode_1'
      - 'Linode_2' 
   - 2 NodeBalancers (without an assigned Firewall)
      - 'NodeBalancer_1'
      - 'NodeBalancer_2'


### Verification steps 
- Make sure the commit history looks accurate
- Scan code and make sure its the right set of code, since the code has been reviewed before, we are more concerned that this PR encapsulates everything needed.
- Run through the feature as a user, you can use the following workflow as a reference:
   - Throughout this workflow, ensure that:
      - There are no grammatical errors
      - The Linode, NodeBalancer and Firewall links bring you to the right page
      - The action is being added to the events: http://localhost:3000/events
   - If you do not have 2 unassigned Linodes and 2 unassigned NodeBalancers create them here: [Linode](http://localhost:3000/linodes/create), [NodeBalancer](http://localhost:3000/nodebalancers/create)
   - First, create a new Firewall (lets call this 'Firewall_1') without assigning any services: http://localhost:3000/firewalls/create
   - Navigate to Firewall_1's Linodes details page - `/firewalls/[firewall_1_id]/linodes`
   - While doing the below steps, make sure the number to the right of the tab headers (ex. Linode<ins>(#)</ins> and NodeBalancer<ins>(#)</ins>) get updated accordingly
   - Assign a Linode (lets call this 'Linode_1') to Firewall_1 - `/firewalls/[firewall_1_id]/linodes/add`
   - Ensure the Linode's table gets populated with the new Linode_1 - `/firewalls/[firewall_1_id]/linodes`
   - Assign another Linode (lets call this 'Linode_2') to Firewall_1 - `/firewalls/[firewall_1_id]/linodes/add`
   - Test the table's search bar functionality at `/firewalls/[firewall_1_id]/linodes` with the two Linodes
   - Unassign Linode_2 using the `remove` button in the table, and ensure the Linode gets properly unassigned.
   - Navigate to Firewall_1's NodeBalancer's details page - `/firewalls/[firewall_1_id]/linodes`
   - Assign a NodeBalancer_1 to Firewall_1 - `/firewalls/[firewall_1_id]/nodebalancers/add`
   - Ensure the NodeBalancer's table get populated with the new NodeBalancer_1 - `/firewalls/[firewall_1_id]/nodebalancers`
   - Ensure NodeBalancer_1's Details page gets updated with Firewall_1 - `/nodebalancers/[nodebalancer_1_id]/networking` and `/nodebalancers/[nodebalancer_1_id]/settings`
   - Navigate back to `/firewalls/[firewall_1_id]/nodebalancers/add` to assign another NodeBalancer, NodeBalancer_2, to Firewall_1
   - Test the NodeBalancer's table's search bar functionality at `/firewalls/[firewall_1_id]/nodebalancers` with the two assigned NodeBalancers.
   -  Unassign NodeBalancer_2 using the `remove` button in the table, and ensure the NodeBalancer gets properly unassigned.
   - Create a new Firewall (lets call this 'Firewall_2') - http://localhost:3000/firewalls/create
   - Ensure you cannot assign the already assigned Linode_1 to Firewall_2 - `/firewalls/[firewall_2_id]/linodes/add` 
      - There is an open [PR](https://github.com/linode/manager/pull/9935) that filters out already assigned services in the dropdowns; however, if this is not mergd by the time you do the review, atleast make sure there is an error that occurs if you do try to add an already assigned service
   - But also ensure that you _can_ assign Linode_2 to Firewall_2, and make sure to assign it
   - Ensure you cannot assign the already assigned NodeBalancer_1 to Firewall_2 - `/firewalls/[firewall_2_id]/nodebalancers/add` 
   - But also ensure that you _can_ assign NodeBalancer_2 to Firewall_2, and make sure to assign it
   - Navigate back to the firewalls landing page - http://localhost:3000/firewalls
   - Ensure that there are two Firewalls in the table - 'Firewall_1' and 'Firewall_2'
   - Ensure that there is a 'Services' column
   - Ensure that each Firewall's Services' column has the corresponding Linode and NodeBalancer that was assigned to. that firewall: Firewall_1 --> Linode_1 and NodeBalancer_1, Firewall_2 --> Linode_2 and NodeBalancer_2
   - Ensure that swapping on and off the feature flag disables and enables all of the feature related ui found [here](https://github.com/linode/manager/pull/9931)
